torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -600,34 +625,49 @@ def moe_forward(self, x, topk_ids, topk_weight):
                 gatherd_idxs[s : s + k] = i % self.experts_per_rank
                 s += k
 
-        # Take necessary space from `token_send_buf` symm mem because we are
-        # going to send them out after expert processing
-        processed_tokens = MoE.token_send_buf[:received]
+        # Prepare buffer for tokens processed by experts
+        if self.shuffle_method == "symm_mem":
+            # Take necessary space from `token_send_buf` symm mem because we are
+            # going to send them out after expert processing
+            processed_tokens = MoE.token_send_buf[: gathered_tokens.shape[0]]
+        else:  # "torch_all_to_all"
+            processed_tokens = torch.empty_like(gathered_tokens)
+
+        # This part processes the tokens routed to the local experts.
+        # TODO: can we use group GEMM here?
         for i, expert in enumerate(self.experts.values()):
             processed_tokens[gatherd_idxs == i] = expert(
                 gathered_tokens[gatherd_idxs == i]
             )
 
-        # Take necessary space from `token_gather_buf` symm mem to receive processed tokens
-        gathered_tokens = MoE.token_gather_buf[: sorted_tokens_shape[0]]
-        received_splits = torch.empty_like(MoE.output_splits)  # unused
-        # EP to DP token shuffle
-        on_device_all_to_all_v(
-            gathered_tokens,
-            received_splits,  # unused
-            processed_tokens,
-            MoE.output_splits,
-            self.ep_group,
-        )
+        # Now shuffle the tokens back to their original owner, i.e. EP to DP shuffle.
+        # The input/output splits are just a reverse of the previous shuffle.
+        if self.shuffle_method == "symm_mem":
+            # Take necessary space from `token_gather_buf` symm mem to receive processed tokens
+            returned_tokens = MoE.token_gather_buf[: sorted_tokens_shape[0]]
+            on_device_all_to_all_v(
+                returned_tokens,
+                MoE.input_splits,  # unused
+                processed_tokens,
+                MoE.output_splits,
+                self.ep_group,
+            )
+        else:  # "torch_all_to_all"
+            returned_tokens = all_to_all_single_autograd(
+                processed_tokens,
+                input_splits.tolist(),
+                output_splits.tolist(),
+                self.ep_group,
+            )
 
-        returned_tokens = torch.empty_like(gathered_tokens)
-        returned_tokens[idxs] = gathered_tokens
+        output_tokens = torch.empty_like(returned_tokens)
+        output_tokens[idxs] = returned_tokens
```

Comment:
approx, in here we can add an un-permute kernel to handle and speedup the returned token re-ordering.